packages/shared/useMediaMatch.ts

```diff
@@ -0,0 +1,37 @@
+import { useEffect, useState } from 'react';
+
+/**
+ * useMediaMatch
+ *
+ * A react hook that signals whether or not a media query is matched.
+ *
+ * @param query The media query to signal on. Example, `"print"` will signal
+ * `true` when previewing in print mode, and `false` otherwise.
+ * @returns Whether or not the media query is currently matched.
+ */
+function useMediaMatch(query: string): boolean {
+  if (typeof window === 'undefined') {
+    console.warn('useMediaMatch cannot function as window is undefined.');
+    return false;
+  }
+
+  const [matchMedia, setMatchMedia] = useState(() => window.matchMedia(query));
+  const [matches, setMatches] = useState(() => matchMedia.matches);
+
+  useEffect(() => {
+    const mm = window.matchMedia(query);
+    setMatchMedia(mm);
+    setMatches(mm.matches);
+  }, [query]);
+
+  useEffect(() => {
+    const listener = (ev: MediaQueryListEventMap['change']) =>
+      setMatches(ev.matches);
+    matchMedia.addEventListener('change', listener);
+    return () => matchMedia.removeEventListener('change', listener);
+  }, [matchMedia]);
+
+  return matches;
+}
```

Comment:
Hmm if I understand the use cases correctly, I don't think you need to setup `matchMedia` as a state:
```suggestion
  const matchMedia = window.matchMedia(query);
  const [matches, setMatches] = useState(() => matchMedia.matches);

  useEffect(() => {
    const mm = window.matchMedia(query);
    setMatches(mm.matches);
  }, [query]);

  useEffect(() => {
    const listener = (ev: MediaQueryListEventMap['change']) =>
      setMatches(ev.matches);
    matchMedia.addEventListener('change', listener);
    return () => matchMedia.removeEventListener('change', listener);
  }, [matchMedia]);

  return matches;
}
```

Reply:
My understanding is that this hook function gets called potentially a lot. I assume running `window.matchMedia` could have some cost to it. For example it resolves a result set that has `matches: boolean`, so the browser would be actively running some match logic potentially very often. I thought it would be best to only instantiate a `matchMedia` instance when the query string changes rather than in a potentially tight loop.

Do you still think we should make this change?

Reply:
Also in your version of the code, you're reacting on dependency `[matchMedia]` in the second useEffect hook which will now change every re-render with your suggested change.

Reply:
@mrvisser Any chance we can do it with just a `useMemo` instead of `useState`? Sorry I joined this discussion a touch late.

Reply:
Oh yea, I think so and it will be a lot cleaner. Will follow up on that.